### PR TITLE
Further energy related work.

### PR DESCRIPTION
**Pull request by jagwire**
Energy calculations in energyBar.js taking into account Entity creation, deletion, and modification.

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
An admin must verify this patch for the pull request build to fire. Type "ok to test" to begin testing this pull request or "add to whitelist" to test this and all future pull requests from this collaborator.

**Comment by PhilipRosedale** [comment]
ok to test

**Comment by hifi-gustavo** [comment]
Build console output for commit: 6c82fce2cd37ec8345813ad2d274384378e18944 available at: https://gist.github.com/bc98499847d73ad4a63f<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR6945.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR6945.dmg

File: interface/src/avatar/MyAvatar.cpp [hunk 1]
```diff
@@ -1764,6 +1776,29 @@ void MyAvatar::lateUpdatePalms() {
     Avatar::updatePalms();
 }
 
+float MyAvatar::getAccelerationEnergy() {
+    glm::vec3 velocity = getVelocity();
+    glm::vec3 changeInVelocity = velocity - priorVelocity;
+    float changeInEnergy = velocity.length()*changeInVelocity.length()*AVATAR_MOVEMENT_ENERGY_CONSTANT;
```
Comment (PhilipRosedale) [inline interface/src/avatar/MyAvatar.cpp]: coding standard:  Put spaces on each side of math operators

File: interface/src/avatar/MyAvatar.cpp [hunk 2]
```diff
@@ -1764,6 +1776,29 @@ void MyAvatar::lateUpdatePalms() {
...
+    
+    return changeInEnergy;
+}
+
+float MyAvatar::getEnergy() {
+    return currentEnergy;
+}
+
+float MyAvatar::getAudioEnergy() {
+    return getAudioLoudness()*AUDIO_ENERGY_CONSTANT;
```
Comment (PhilipRosedale) [inline interface/src/avatar/MyAvatar.cpp]: space

File: libraries/entities/src/EntityScriptingInterface.cpp [hunk 3]
```diff
@@ -119,11 +118,26 @@ EntityItemProperties convertLocationFromScriptSemantics(const EntityItemProperti
...
+    //check our user-defined script, if it exists, to see if we have the energy to do what we want.
+    auto dimensions = propertiesWithSimID.getDimensions();
+    float volume = dimensions.x * dimensions.y * dimensions.z;
+    double cost = calculateCost(propertiesWithSimID.getDensity()*volume,
+                                0, // 0 - old velocity because the entity doesn't exist yet
+                                propertiesWithSimID.getVelocity().length()); //new velocity as specified by scripter
+    
+    //compare cost with MyAvatar.getEnergy();
+    if(cost > _currentAvatarEnergy) {
+        return id;
```
Comment (PhilipRosedale) [inline libraries/entities/src/EntityScriptingInterface.cpp]: is this the right return value for a fail?  should we return something to indicate the add failed for lack of energy?

File: libraries/entities/src/EntityScriptingInterface.cpp [hunk 4]
```diff
@@ -212,11 +229,35 @@ EntityItemProperties EntityScriptingInterface::getEntityProperties(QUuid identit
...
         return id;
     }
     // If we have a local entity tree set, then also update it.
+    
+    EntityItemPointer entity = _entityTree->findEntityByEntityItemID(entityID);
+    
+    auto dimensions = entity->getDimensions();
+    float density = entity->getDensity();
+    float volume = dimensions.x * dimensions.y * dimensions.z;
+    double cost = calculateCost(density*volume,
```
Comment (PhilipRosedale) [inline libraries/entities/src/EntityScriptingInterface.cpp]: spaces

File: libraries/entities/src/EntityScriptingInterface.cpp [hunk 5]
```diff
@@ -315,6 +356,18 @@ void EntityScriptingInterface::deleteEntity(QUuid id) {
     if (_entityTree) {
         _entityTree->withWriteLock([&] {
             EntityItemPointer entity = _entityTree->findEntityByEntityItemID(entityID);
+            
+            auto dimensions = entity->getDimensions();
+            float volume = dimensions.x * dimensions.y * dimensions.z;
+            float density = entity->getDensity();
+            double cost = calculateCost(density*volume,
```
Comment (PhilipRosedale) [inline libraries/entities/src/EntityScriptingInterface.cpp]: spaces

File: libraries/entities/src/EntityScriptingInterface.cpp [hunk 6]
```diff
@@ -212,11 +229,35 @@ EntityItemProperties EntityScriptingInterface::getEntityProperties(QUuid identit
...
+    
     if (!_entityTree) {
         queueEntityMessage(PacketType::EntityEdit, entityID, properties);
         return id;
     }
     // If we have a local entity tree set, then also update it.
+    
+    EntityItemPointer entity = _entityTree->findEntityByEntityItemID(entityID);
+    
+    auto dimensions = entity->getDimensions();
```
Comment (PhilipRosedale) [inline libraries/entities/src/EntityScriptingInterface.cpp]: When I ran this PR, distant grabbing an object crashed Interface at this line.

**Comment by imgntn** [comment]
test this please

**Comment by hifi-gustavo** [comment]
Build console output for commit: 6c82fce2cd37ec8345813ad2d274384378e18944 available at: https://gist.github.com/bc98499847d73ad4a63f<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR6945.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR6945.dmg

File: interface/src/avatar/MyAvatar.cpp (hunk 1 above)
Comment (PhilipRosedale) [inline interface/src/avatar/MyAvatar.cpp]: coding standard: one space on each side of the '*'

**Pull request closed without merging**